### PR TITLE
Support configuring request metadata for Ray client gRPC

### LINKED ISSUE #22945: [core][Feature] Ability to set Headers for gRPC channels.
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Currently gRPC supports setting custom headers prior to creating the channels. However, Ray.init() doesn't support custom headers as parameters. This can be limiting due to multiple reasons: proxy authentication, etc. 

Is there a workaround to set a custom header for the gRPC channel? We do it for client_id, and the underlying gRPC SDK support it: See [here.](https://github.com/ray-project/ray/blob/2302b4eea87c480699797e4bf770d46d0b758379/python/ray/util/client/worker.py#L98)

### Use case

Need the ability to pass custom headers through the gRPC channels. This option will solve a variety of different use cases.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by peterghaddad**
@mwtian 

## Why are these changes needed?

These changes are needed to enable the end-user to pass headers for authenticating through gRPC. This enhances enterprise support of Ray.

## Related issue number

https://github.com/ray-project/ray/issues/22945

Closes #22945

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/worker.py
Comment (ckw017) [inline python/ray/worker.py:810]: PR looks good overall. Can we make this a bit more specific, i.e. "List of tuples representing request metadata for Ray Client gRPC."

**Review by ckw017 — APPROVED** [review]

**Comment by ckw017** [comment]
Needs a merge with master to resolve conflict

File: python/ray/worker.py
Comment (peterghaddad) [reply python/ray/worker.py]: Addressed!

**Comment by peterghaddad** [comment]
@ckw017 I merged master into my branch. I am no longer seeing any merge conflicts locally, but seeing that Github is saying there are some conflicts in worker.py?

**Comment by ckw017** [comment]
Ah hmm, can you try `git pull upstream master`? If that doesn't work can you share what `git remote -v` outputs

**Comment by ckw017** [comment]
Yup looks like that worked, let's wait for tests to rerun.

**Comment by peterghaddad** [comment]
@ckw017 Thank you for all the review!

File: python/ray/_private/worker.py
Comment (ckw017) [inline python/ray/_private/worker.py:1095]: Looks like the [lint](https://buildkite.com/ray-project/ray-builders-pr/builds/36780#0181a754-2fb9-4014-8df3-3903139bfac3) is failing on here, can you remove the `(list)` part

Comment (peterghaddad) [reply python/ray/_private/worker.py]: @ckw017  Sure thing! Resolved. Sorry, not used to CI in Github.

File: python/ray/_private/worker.py
Comment (mwtian) [inline python/ray/_private/worker.py:1003]: We probably want to implement the underlying functionality first, before making the option available in the public API. Also,
- `metadata` sounds a bit too generic. How about `request_headers`?
- We need a way to pass this to `ray start` command, in `scripts.py`

Comment (peterghaddad) [reply python/ray/_private/worker.py:1003]: @mwtian Can we create an additional issue for appending this to the ray start command? This has to do with the SDK and gRPC client specifically, so uncertain why it needs to be added with ray start.

Comment (mwtian) [reply python/ray/_private/worker.py:1003]: I see. `ray.init()` is a public API. We probably need an approval to add a parameter.

Since this is going to be Ray client specific, how about we use an environment variable like `RAY_CLIENT_GRPC_HEADER` to pass the information?

cc @scv119 @pcmoritz @ericl @zhe-thoughts

**Comment by peterghaddad** [comment]
@mwtian They do not! It is the initial connection to the head node in this case. From my work on Ray Clusters, communication at the Cluster level seem to be handled well!

**Comment by mwtian** [comment]
Sounds good. I think we only need another pair of eyes on the new option then.

**Comment by ckw017** [comment]
If we don't want to commit to changing the API right now, we can also document it as `_metadata` (or `_client_metadata` if we want to be more specific) and treat as an experimental argument for now. At the moment this is how [`_credentials`](https://github.com/ray-project/ray/blob/21760fd3bab4723c92788a3b82385370c0990f3d/python/ray/client_builder.py#L213-L215) is currently passed.

**Comment by peterghaddad** [comment]
@ckw017 Sounds like a plan! I made a push to address this. Is this good to go?

**Comment by mwtian** [comment]
I would prefer using an env var like `RAY_CLIENT_...`. `_metadata` just seems too generic as a parameter in `ray.init()`. But others may have different opinions. Anyway I'm not able to merge this on my own. @ckw017 can you find a Ray committer to review and help push this PR through?
cc @scv119 @pcmoritz @ericl @zhe-thoughts

**Comment by peterghaddad** [comment]
Thanks for the reply, @mwtian. Just so we're all on the same page: Metadata is already a parameter in `ray.connect()`, and this fix essentially just brings it up a level to the `ray.init` function since ray.init eventually calls `ray.connect().`

For example:
```ray.connect(metadata = _metadata)```

 If the user wants to pass additional metadata, then they would do so through the dictionary i.e if the Cluster requires additional data.  I don't think it would make sense to rename it to something else without changing the parameter in `ray.connect()` also.

**Comment by mwtian** [comment]
Yeah, but `ray.connect()` is a Ray client only API, so `metadata` there is less ambiguous. `ray.init()` is an API used by both Ray client and Ray cluster.

**Comment by peterghaddad** [comment]
@mwtian Ah okay, thanks! I'm okay with using the environment variable; however, I think having the ability to pass a dictionary as metadata is the way to go, and i'm uncertain if there is a clean way of doing this with an environment variable. If the Ray cluster uses gRPC for connections, then I think using `_metadata` would be okay since the context doesn't change.

**Comment by peterghaddad** [comment]
Hi @ckw017, any updates on this PR? Can we get this merged in? Thanks!

**Review by ijrsvt — COMMENTED** [review]
Rather than changing the signature of `ray.init`, can we follow the pattern of  how `_credentials` is passed:

Something like the following would get `_metadata` to the `connect` call. 
```
diff --git a/python/ray/client_builder.py b/python/ray/client_builder.py
index 51ccffd38c..a25db0ad6e 100644
--- a/python/ray/client_builder.py
+++ b/python/ray/client_builder.py
@@ -116,6 +116,7 @@ class ClientBuilder:
         # Set to False if ClientBuilder is being constructed by internal
         # methods
         self._deprecation_warn_enabled = True
+        self._metadata = None
 
     def env(self, env: Dict[str, Any]) -> "ClientBuilder":
         """
@@ -179,6 +180,7 @@ class ClientBuilder:
             job_config=self._job_config,
             _credentials=self._credentials,
             ray_init_kwargs=self._remote_init_kwargs,
+            metadata=self._metadata,
         )
         get_dashboard_url = ray.remote(ray._private.worker.get_dashboard_url)
         dashboard_url = ray.get(get_dashboard_url.options(num_cpus=0).remote())
@@ -229,6 +231,10 @@ class ClientBuilder:
             self._credentials = kwargs["_credentials"]
             del kwargs["_credentials"]
 
+        if "_metadata" in kwargs.keys():
+            self._metadata = kwargs["_metadata"]
+            del kwargs["_metadata"]
+
         if kwargs:
             expected_sig = inspect.signature(ray_driver_init)
             extra_args = set(kwargs.keys()).difference(expected_sig.parameters.keys())
```

**Comment by peterghaddad** [comment]
@ijrsvt I just pushed some changes to address these comments. Can you take a look, and ensure this looks okay?

Looks like there is a @deprecated annotation on-top of the function. Will we be able to still pass _metadata as part of ray.init()? i.e `ray.init(_metadata)`.

**Comment by ijrsvt** [comment]
@peterghaddad The deprecated annotation on `python/ray/util/client_connect.py::connect` is to discourage something like `ray.util.client_connect.connect()` in user code.

File: python/ray/util/client_connect.py
Comment (ijrsvt) [inline python/ray/util/client_connect.py:18]: Final nit, can we leave this signature the same?

Comment (peterghaddad) [reply python/ray/util/client_connect.py]: @ijrsvt 
Of course! I changed `connect` to take metadata.

Just to make sure this looks okay:

clinet_builder.py
```python
        client_info_dict = ray.util.client_connect.connect(
            self.address,
            job_config=self._job_config,
            _credentials=self._credentials,
            ray_init_kwargs=self._remote_init_kwargs,
            metadata=self._metadata,
        )
```

Do we think this looks good to go?

**Review by ijrsvt — APPROVED** [review]
LGTM now!

**Comment by peterghaddad** [comment]
@ijrsvt Thanks! Do we need more approvals to get this Merged? First time contributing to Ray!

**Comment by ijrsvt** [comment]
Can you rebase/merge master?

**Comment by peterghaddad** [comment]
@ijrsvt Sure thing. Finished!

**Comment by peterghaddad** [comment]
@ijrsvt @ckw017 Anything blocking us from merging this in? Looks like all checks passed!

**Pull request merged**
